pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/groupby/DictionaryBasedGroupKeyGenerator.java

```diff
@@ -106,18 +108,26 @@ public DictionaryBasedGroupKeyGenerator(TransformOperator transformOperator,
 
       _isSingleValueColumn[i] = transformOperator.getResultMetadata(groupByExpression).isSingleValue();
     }
-
     if (longOverflow) {
       _globalGroupIdUpperBound = numGroupsLimit;
-      _rawKeyHolder = new ArrayMapBasedHolder(_globalGroupIdUpperBound);
+      if (!mapBasedRawKeyHolders.containsKey(ArrayMapBasedHolder.class.getName())) {
```

Comment:
Use `computeIfAbsent()` to save one extra map lookup.
Also, because you already know the type of key holders, maybe add a separate class to store different type of holders without map lookup?

Reply:
those types are internal and not exposed outside `DictionaryBasedGroupKeyGenerator.java`